### PR TITLE
Add capability to have configurable aggregation temporality for OTLP Registry

### LINKED ISSUE #3145: Configurable aggregation temporality for OTLP registry
**Please describe the feature request.**
OTLP format supports delta and cumulative temporality. Currently the OTLP registry does not make this configurable and cumulative is always sent. While this should work for compliant OTLP receivers, as far as I understand, some users may have a preference for one and we could support making this configurable.

**Rationale**
OTLP explicitly does not have a preference of aggregation temporality. Since OTLP is not tied to a specific metrics backend, there is less reason for us to use one temporality over the other.

**Additional context**
See initial discussion: https://github.com/micrometer-metrics/micrometer/pull/3129#discussion_r850680118. There is prior art in that the `SimpleMeterRegistry` allows configuring whether it is cumulative or step based.
/cc @twicksell

### PR DESCRIPTION
**Pull request by lenin-jaganathan**
OTLP format supports delta and cumulative temporality. But currently in micrometer, the support for OTLP Metrics with Delta Temporality is not available. This PR adds the capability to export meters with delta temporality(https://github.com/micrometer-metrics/micrometer/issues/3145). Below are some behaviors this PR includes
* capability to configure aggregation Temporality via property `otlp.aggregationTemporality`
* Delta Temporality is supported by using StepMeters wherever applicable
* Capability to export max as part of the HistogramDataPoint for Delta Temporality

Implementation Information:
When Delta Temporality is enabled the following applies,
* start time(`StartTimeUnixNano`) for the meter will be set to the start of the previous step.
* end time(`TimeUnixNano `) for the meter will be set to the end of the previous step or the step for which data is being reported.

(the previous step - indicates the last fully completed step.)

TODO:
- [x] Add tests for the new features added
- [ ] Make the support for Delta Histogram using a StepHistogram(basically the count, total should match with the value recorded in Histogram) as mentioned in the OTLP protocol
- [ ] Possibly rename some of the classes and minimize code duplications.

fixes #3145

### CODE REVIEW

File: implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpStepDistributionSummary.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+ * limitations under the License.
+ */
+package io.micrometer.registry.otlp;
+
+import io.micrometer.common.lang.Nullable;
+import io.micrometer.core.instrument.Clock;
+import io.micrometer.core.instrument.distribution.*;
+import io.micrometer.core.instrument.step.StepDistributionSummary;
+
+class OtlpStepDistributionSummary extends StepDistributionSummary {
```
Comment (jonatan-ivanov) [inline implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpStepDistributionSummary.java:23]: I'm not 100% sure that this is a good idea but this class seem the same as `OtlpCumulativeDistributionSummary` except:
1. `TimeWindowFixedBoundaryHistogram`/`DistributionStatisticConfig`/etc. are created differently
2. Cumulative has `getStartTimeNanos`

So I was thinking if we should have the common logic in a common place, e.g.:
1. Having only one `OtlpDistributionSummary` that can handle both (`getStartTimeNanos` will be there in the step case too).
2. Having a common parent/super class that cumulative/step can extend and provide the `TimeWindowFixedBoundaryHistogram`/`DistributionStatisticConfig`/etc. for example in ctor.

File: implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpStepTimer.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2022 VMware, Inc.
```
Comment (jonatan-ivanov) [inline implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpStepTimer.java:2]: Could you please fix copyright headers (2023)?

File: implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpStepDistributionSummary.java (hunk 1 above)
Comment (jonatan-ivanov) [reply implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpStepDistributionSummary.java:23]: I think `OtlpStepTimer` has the same situation.

**Review by jonatan-ivanov — CHANGES_REQUESTED** [review]

File: implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpConfig.java [hunk 3]
```diff
@@ -87,6 +87,19 @@ default Map<String, String> resourceAttributes() {
...
 
+    /**
+     * {@link AggregationTemporality} of the OtlpMeterRegistry. This determines whether
+     * the meters should be cumulative(AGGREGATION_TEMPORALITY_CUMULATIVE) or
+     * step(AGGREGATION_TEMPORALITY_DELTA).
+     * @return the aggregationTemporality for OtlpRegistry
+     * @see <a href=
+     * "https://opentelemetry.io/docs/reference/specification/metrics/data-model/#temporality">Temporality</a>
+     */
+    default int getAggregationTemporality() {
```
[the remark this answers is not in the document]
Comment (lenin-jaganathan) [reply implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpConfig.java:98]: I started with an enum's and eventually moved back to using constants in `AggregationTemporality`. I don't have a preference here but somehow felt an integer might be appropriate I leave this to you guys on determining the best way. I was also seeing this from other integrations' points of view like spring-boot-actuator. As I said I will leave this to you guys and not be determined by our use case in the larger interest.

File: implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpConfig.java [hunk 4]
```diff
@@ -87,6 +87,19 @@ default Map<String, String> resourceAttributes() {
...
+        return getInteger(this, "aggregationTemporality")
+                .orElse(AggregationTemporality.AGGREGATION_TEMPORALITY_CUMULATIVE_VALUE);
```
[the remark this answers is not in the document]
Comment (lenin-jaganathan) [reply implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpConfig.java:100]: https://opentelemetry.io/docs/reference/specification/sdk-environment-variables/#metrics-sdk-configuration - This does not specify anything for aggregation temporality. 
On another note, I believe the micrometer already does not honor properties like "OTEL_METRIC_EXPORT_INTERVAL" instead we go with the step property. So, probably we can think about how this bridge with OTLP env variables can be established.

File: implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMeterRegistry.java [hunk 5]
```diff
@@ -244,26 +262,39 @@ Metric writeHistogramSupport(HistogramSupport histogramSupport) {
...
+        return config.getAggregationTemporality() == AggregationTemporality.AGGREGATION_TEMPORALITY_DELTA_VALUE;
+    }
+
+    private long getStartTimeNanos(Meter meter) {
+        return isDeltaAggregationTemporality() ? getEndTimeNanos() - config.step().toNanos()
+                : ((StartTimeAwareMeter) meter).getStartTimeNanos();
+    }
+
+    private long getEndTimeNanos() {
+        return isDeltaAggregationTemporality() ? (clock.wallTime() / config.step().toMillis()) * config.step().toNanos()
```
[the remark this answers is not in the document]
Comment (lenin-jaganathan) [reply implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMeterRegistry.java:294]: Going to work more on this and the tests too. Just wanted to kick start this PR with an initiative for Delta-based Aggregation Temporality.
But the overall idea is not to use different times for Step times. What I tried to achieve is to have normalized steps like this - 00:00:01.000 - 00:00:02.00, 00:00:02.000 - 00:00:03.00. This will make sure all the exported meters for a given step will have consistent timing.

File: implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMeterRegistry.java [hunk 6]
```diff
@@ -244,26 +262,39 @@ Metric writeHistogramSupport(HistogramSupport histogramSupport) {
...
+                        .setStartTimeUnixNano(getStartTimeNanos((functionTimer))).setTimeUnixNano(getEndTimeNanos())
                         .setSum(functionTimer.totalTime(getBaseTimeUnit())).setCount((long) functionTimer.count())))
                 .build();
     }
 
+    private boolean isDeltaAggregationTemporality() {
+        return config.getAggregationTemporality() == AggregationTemporality.AGGREGATION_TEMPORALITY_DELTA_VALUE;
+    }
+
+    private long getStartTimeNanos(Meter meter) {
```
[the remark this answers is not in the document]
Comment (lenin-jaganathan) [reply implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMeterRegistry.java:288]: I agree from the calculation point of view that we will have 2 calls to "clock.wallTime()". But since we are trying to normalize the steps, the final values would have much difference and will point to the same step start and end times.

File: implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMeterRegistry.java [hunk 7]
```diff
@@ -236,6 +251,9 @@ Metric writeHistogramSupport(HistogramSupport histogramSupport) {
         HistogramDataPoint.Builder histogramDataPoint = HistogramDataPoint.newBuilder().addAllAttributes(tags)
                 .setStartTimeUnixNano(startTimeNanos).setTimeUnixNano(wallTimeNanos).setSum(total).setCount(count);
 
+        if (isDeltaAggregationTemporality()) {
```
[the remark this answers is not in the document]
Comment (lenin-jaganathan) [reply implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMeterRegistry.java:254]: Haha! More about this is being discussed in this [issue](https://github.com/micrometer-metrics/micrometer/issues/3144). There is a discussion about what max should imply for a Cumulative implementation. See this OTLP [proto](https://github.com/jmacd/opentelemetry-proto/blob/2f9dad40420e6585680861a8bd70135cff77a113/opentelemetry/proto/metrics/v1/metrics.proto#L496-L500) where it clearly says min and max should be min and max over start_time and end_time which is not so useful. So, that I believe could be its own PR with more thoughts on what max is and more importantly checking with the OTLP community on this.

File: implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpStepDistributionSummary.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+     * Create a new {@code StepDistributionSummary}.
+     * @param id ID
+     * @param clock clock
+     * @param distributionStatisticConfig distribution static configuration
+     * @param scale scale
+     * @param stepMillis step in milliseconds
+     */
+    public OtlpStepDistributionSummary(Id id, Clock clock, DistributionStatisticConfig distributionStatisticConfig,
+            double scale, long stepMillis) {
+        super(id, clock, DistributionStatisticConfig.builder().percentilesHistogram(false).serviceLevelObjectives()
```
[the remark this answers is not in the document]
Comment (lenin-jaganathan) [reply implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpStepDistributionSummary.java:38]: I am already planning to abstract some of this and avoid code duplication will consider this as well.

File: implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpStepDistributionSummary.java (hunk 1 above)
Comment (lenin-jaganathan) [reply implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpStepDistributionSummary.java:23]: That's right see one of my TODOs in the PR description. This was one of my POC codes and it needs some re-designing and better code re-organization. Will start on that part and welcome any suggestions there.

File: implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpConfig.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (shakuzen) [reply implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpConfig.java:98]: An enum would be better. We can decode strings to enums and a string of `cumulative` or `delta` will be much more understandable for people writing or reading configuration.

File: implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMeterRegistry.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (jonatan-ivanov) [reply implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMeterRegistry.java:285]: Yeah, these are the things we ned to think through and also test since we can bump into corner cases/tricky scenarios, right now, it makes sense to me but haven't tried to do anything "mean" to it. :D

File: implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMeterRegistry.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (jonatan-ivanov) [reply implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMeterRegistry.java]: Oh right, I remember. It is not really the concern of this PR anyway, but just for the record I agree with Tommy on this:
>I think from a usefulness perspective, when using cumulative temporality, we should publish our time-windowed max as a separate gauge and not set max on the histogram.

File: implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpStepDistributionSummary.java (hunk 1 above)
Comment (jonatan-ivanov) [reply implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpStepDistributionSummary.java:23]: Feel free to ping us if you need some help.

Comment (lenin-jaganathan) [reply implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpStepDistributionSummary.java:23]: Yeah. I might need some especially in restructuring some of these classes. Already, OtlpMeterRegisrty does what I believe was not How registries were meant to work doing 2 things - Step and Cumulative. And the same goes with the Timers and DistributionSummaries because there are different things inherently - Step and Cumulative. So, any help on this would be helpful.

File: implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpConfig.java [hunk 9]
```diff
@@ -87,6 +87,19 @@ default Map<String, String> resourceAttributes() {
...
+     * {@link AggregationTemporality} of the OtlpMeterRegistry. This determines whether
+     * the meters should be cumulative(AGGREGATION_TEMPORALITY_CUMULATIVE) or
+     * step(AGGREGATION_TEMPORALITY_DELTA).
+     * @return the aggregationTemporality for OtlpRegistry
+     * @see <a href=
+     * "https://opentelemetry.io/docs/reference/specification/metrics/data-model/#temporality">Temporality</a>
+     */
+    default AggregationTemporality aggregationTemporality() {
+        return getEnum(this, AggregationTemporality.class, "aggregationTemporality")
+                .orElse(AggregationTemporality.AGGREGATION_TEMPORALITY_CUMULATIVE);
```
Comment (shakuzen) [inline implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpConfig.java:100]: It seems like it may be better to create our own enum that we can map to the OTLP proto enum (if need be) rather than directly referencing it here. Their enum has 4 values but I think we probably only want to allow cumulative or delta for now. What would we do if this configuration is set to `UNRECOGNIZED`? Does it make sense to allow configuring `AGGREGATION_TEMPORALITY_UNSPECIFIED`?

Comment (lenin-jaganathan) [reply implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpConfig.java:100]: [OTLP Proto](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L247) says that,

> UNSPECIFIED is the default AggregationTemporality, it MUST not be used.
Honestly, not sure how to interpret it.  But, I assume that clients should not(it says MUST not🙂) use that???

Coming to having our own Enum seems not a bad idea. It allows external frameworks to avoid introducing additional dependencies.

File: implementations/micrometer-registry-otlp/src/test/java/io/micrometer/registry/otlp/DeltaOtlpMeterRegistryTest.java [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+        assertHistogram(publishTimeAwareWrite(timer), TimeUnit.MINUTES.toNanos(4), TimeUnit.MINUTES.toNanos(5),
+                "milliseconds", 0, 0, 0);
+        timer.record(1, TimeUnit.MILLISECONDS);
+        this.stepOverNStep(1);
+        assertHistogram(publishTimeAwareWrite(timer), TimeUnit.MINUTES.toNanos(5), TimeUnit.MINUTES.toNanos(6),
+                "milliseconds", 1, 1, 1);
+    }
+
+    @Test
+    void timerWithHistogram() {
```
Comment (lenin-jaganathan) [inline implementations/micrometer-registry-otlp/src/test/java/io/micrometer/registry/otlp/DeltaOtlpMeterRegistryTest.java:123]: @shakuzen / @jonatan-ivanov I believe you guys are already aware that the TImeWindowFixedBoundaryHistogram will have very poor accuracy for StepBased registries. 

To Summarise:
The issue with that is it uses a sliding bucket which creates 3 buckets(assumption, step=1min, the default buffer length is 3). And these buckets are rotated every 20 seconds (step/buffer length) which makes the histogram always error-prone I,e it will always undercount the values. This is the reason we used [DeltaHistogramCounts](https://github.com/micrometer-metrics/micrometer/blob/main/implementations/micrometer-registry-signalfx/src/main/java/io/micrometer/signalfx/DeltaHistogramCounts.java) implementation in SignalFxMeterRegistry for our histogram use-cases. With OTLP coming into the picture, I think we should at least have a similar implementation (or basic StepBased Histogram) which will give more accurate information. And moreover, the percentiles [don't align with Step Boundaries ](https://github.com/micrometer-metrics/micrometer/issues/3087) which also makes it difficult to correlate them.

File: implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMeterRegistry.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (lenin-jaganathan) [reply implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMeterRegistry.java:299]: Added additional test cases for this.

File: implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMeterRegistry.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (lenin-jaganathan) [reply implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMeterRegistry.java]: Tried to fix this in the lastest commit.

File: implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/AggregationTemporality.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.registry.otlp;
+
+public enum AggregationTemporality {
+
+    AGGREGATION_TEMPORALITY_DELTA, AGGREGATION_TEMPORALITY_CUMULATIVE;
```
Comment (shakuzen) [inline implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/AggregationTemporality.java:20]: I'll take a closer look next week. Just a nit pick for now: I think it'd be better if we named these `DELTA` and `CUMULATIVE` so that configuration is more terse. It should be apparent from the context that these values are for the aggregation temporality.

File: implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpStepDistributionSummary.java (hunk 1 above)
Comment (pirgeo) [reply implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpStepDistributionSummary.java:23]: I want to throw in the idea of get-and-reset semantics for Histograms. Basically, you record the data, snapshot it, and then reset the histogram on every export. This circumvents the whole timer and rotation code, which I am not sure is required for a delta-based export. Just leaving this here as an idea, let me know if there is something I am missing

File: implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMeterRegistry.java [hunk 12]
```diff
@@ -82,13 +90,18 @@ private OtlpMeterRegistry(OtlpConfig config, Clock clock, HttpSender httpSender)
         this.config = config;
         this.httpSender = httpSender;
         this.resource = Resource.newBuilder().addAllAttributes(getResourceAttributes()).build();
+        this.isDeltaAggregationTemporality = config
+                .aggregationTemporality() == AggregationTemporality.AGGREGATION_TEMPORALITY_DELTA;
+        this.otlpAggregationTemporality = AggregationTemporality.mapToOtlp(config.aggregationTemporality());
```
[the remark this answers is not in the document]
Comment (lenin-jaganathan) [reply implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMeterRegistry.java]: I would be more happy to commit to the suggestion so that it reflects the contribution but unfortunately Github doesn't let me commit the suggestion. So, I am going to fix with a new commit.

File: implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMeterRegistry.java [hunk 13]
```diff
@@ -123,19 +136,27 @@ protected <T> Gauge newGauge(Meter.Id id, @Nullable T obj, ToDoubleFunction<T> v
 
     @Override
     protected Counter newCounter(Meter.Id id) {
-        return new OtlpCounter(id, this.clock);
+        return isDeltaAggregationTemporality ? new StepCounter(id, this.clock, config.step().toMillis())
+                : new OtlpCounter(id, this.clock);
```
[the remark this answers is not in the document]
Comment (lenin-jaganathan) [reply implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMeterRegistry.java:139]: I am waiting for @shakuzen / @jonatan-ivanov  to come up with their review for some of the re-factoring parts. I am sure they will come up with a better naming convention than me. 😉

**Comment by jonatan-ivanov** [comment]
Small update: as I was looking into how to do this without code duplication, I thought having some integration tests could be timely, so: https://github.com/micrometer-metrics/micrometer/pull/3668

**Comment by bogdandrutu** [comment]
@jonatan-ivanov what is the missing part to move forward with this PR?

**Comment by lenin-jaganathan** [comment]
@jonatan-ivanov / @shakuzen Is there any update on this?

**Comment by jonatan-ivanov** [comment]
I resoled the conflicts (merged `main` to this and kept the history) and applied formatting changed. I'm planning to look more into this this week.

**Comment by jonatan-ivanov** [comment]
@lenin-jaganathan I did some polishing on the PR, could you please check if they look ok for you? I think this looks good, I want to check the publishing time behavior tomorrow and hopefully integrate it with the integration tests I added earlier.

**Comment by lenin-jaganathan** [comment]
@jonatan-ivanov  Looks good as far as the changes are concerned. But I do have a couple of concerns overall which can be taken as separate issues as well,
* https://github.com/micrometer-metrics/micrometer/issues/1882 will re-surface for the OTLP delta-based registry.
* StepHistogram (or something similar) which will follow the OTLP protocol where the count() should match with the counts in the histogram. It helps in overcoming one of the [issues discussed earlier in this pr](https://github.com/micrometer-metrics/micrometer/pull/3625#discussion_r1115580615).

File: implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpCumulativeDistributionSummary.java [hunk 14]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2022 VMware, Inc.
+ * Copyright 2023 VMware, Inc.
```
[the remark this answers is not in the document]
Comment (lenin-jaganathan) [reply implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpCumulativeDistributionSummary.java:2]: @jonatan-ivanov Can you confirm this?

**Comment by shakuzen** [comment]
> #1882 will re-surface for the OTLP delta-based registry.

I think the `close` implementation from `StepMeterRegistry` will need to be copied to `OtlpMeterRegistry` for that.

**Comment by lenin-jaganathan** [comment]
> I think the close implementation from StepMeterRegistry will need to be copied to OtlpMeterRegistry for that.

I tried that already (though it is a bad code duplication), but the StepMeter / _manualRollover() has package-private scope(which is understandable)🙂

**Comment by jonatan-ivanov** [comment]
I did some further refactoring and added integration tests, most of the things look good to me, but I have two concerns:

1. We are changing the publishing timestamp for cumulative from
```
.setTimeUnixNano(TimeUnit.MILLISECONDS.toNanos(this.clock.wallTime()))
```
to
```
.setTimeUnixNano(publishTimeNanos)
```
In the first case, meters published together can have different timestamps, in the second they can't. E.g.: let's say we have two gages and pulling them can take 5 seconds, in the first case their timestamps will be ~5 seconds apart, in the second case they will be off. I think the original approach is closer to the truth, what do you think?

2. @lenin-jaganathan Did you try this with a real backend? I wrote an integration test with the OTel collector and it seems only `Counter` and `Gauge` works, `Timer` and `DistributionSummary` does not I might miss something.

<details>
<summary>We are sending the data to the collector (click to reveal an example request)</summary>
```
resource_metrics {
  resource {...}
  scope_metrics {
    metrics {
      name: "test.gauge"
      gauge {
        data_points {
          time_unix_nano: 1680302436000000000
          as_double: 12.0
        }
      }
    }
    metrics {
      name: "test.timer"
      unit: "milliseconds"
      histogram {
        data_points {
          start_time_unix_nano: 1680302434000000000
          time_unix_nano: 1680302436000000000
          count: 1
          sum: 123.0
          max: 123.0
        }
        aggregation_temporality: AGGREGATION_TEMPORALITY_DELTA
      }
    }
    metrics {
      name: "test.distributionsummary"
      histogram {
        data_points {
          start_time_unix_nano: 1680302434000000000
          time_unix_nano: 1680302436000000000
          count: 1
          sum: 24.0
          max: 24.0
        }
        aggregation_temporality: AGGREGATION_TEMPORALITY_DELTA
      }
    }
    metrics {
      name: "test.counter"
      sum {
        data_points {
          start_time_unix_nano: 1680302434000000000
          time_unix_nano: 1680302436000000000
          as_double: 42.0
        }
        aggregation_temporality: AGGREGATION_TEMPORALITY_DELTA
        is_monotonic: true
      }
    }
  }
}
```
</details>

But the Prometheus response does not have them:
```
# HELP test_counter 
# TYPE test_counter unknown
test_counter{job="test",service_name="test",telemetry_sdk_language="java",telemetry_sdk_name="io.micrometer"} 42.0
# HELP test_gauge 
# TYPE test_gauge gauge
test_gauge{job="test",service_name="test",telemetry_sdk_language="java",telemetry_sdk_name="io.micrometer"} 12.0
# EOF
```

Additionally, these tests are timing out on CircleCI and I cannot re-trigger them since for some reason they are in @lenin-jaganathan's org. :o

**Comment by lenin-jaganathan** [comment]
1. `We are changing the publishing timestamp for cumulative. ` - I get your point, especially for the cumulative registry. I will have a go at it and get back.

2. I did try my initial changes with a back-end that supports Delta Aggregation. I will have a look at it again, but I don't expect Prometheus to support the Delta Aggregation(https://opentelemetry.io/docs/reference/specification/metrics/sdk_exporters/prometheus/). If what I assume is right, it should only process cumulative aggregation.

**Comment by jonatan-ivanov** [comment]
Yeah, Prometheus not supporting delta is ok (it can't really), but I would have assumed that the OTel collector can accept OTLP with delta and export it in the Prometheus format (as cumulative). :(
Because of this, could you please also revert my last two commits?

**Comment by lenin-jaganathan** [comment]
@jonatan-ivanov I did verify the Histogram datapoint after your comment and I don't see any issues with them. I see the data points in our back end and it just works fine.

Do you want me to revert your last 2 commits?

**Comment by jonatan-ivanov** [comment]
👍🏼 
Yes please if you are restoring the timestamp for cumulative too.
I believe with those the PR should be ready for merging.

File: implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpCumulativeTimer.java [hunk 15]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2022 VMware, Inc.
+ * Copyright 2023 VMware, Inc.
```
Comment (pirgeo) [inline implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpCumulativeTimer.java:2]: Sorry for bothering you with this mundane thing _again_, @jonatan-ivanov, was the year supposed to change, or should it stay the same as it was originally? Or does it not matter at all?

**Review by pirgeo — APPROVED** [review]
I gave it another pass and tried running it with a collector that simply logs what it receives. Looks like it all works to me. I haven't done extensive tests with a backend yet though.

**Review by jonatan-ivanov — APPROVED** [review]

**Comment by jonatan-ivanov** [comment]
@lenin-jaganathan Thank you for the changes!

@shakuzen I think this is in a good enough shape to merge it, what do you think?

File: implementations/micrometer-registry-otlp/src/test/java/io/micrometer/registry/otlp/DeltaOtlpMeterRegistryTest.java (hunk 10 above)
Comment (shakuzen) [reply implementations/micrometer-registry-otlp/src/test/java/io/micrometer/registry/otlp/DeltaOtlpMeterRegistryTest.java:132]: We'll need to change this. SignalFx was the first step-based registry to support percentile histograms, so we hadn't taken close enough look at this in general, I think. The histogram support is indeed not right, which we can fix after merging this, but I don't usually like merging something I know has a bug.

The percentiles and max not aligning with step boundaries was intentional (there's a relevant old issue with the history), but I don't think it makes sense in the OTLP context because the specification is clear about what the meaning of the max or bucket counts is supposed to represent, and it needs to align with the time interval.

File: implementations/micrometer-registry-otlp/src/test/java/io/micrometer/registry/otlp/DeltaOtlpMeterRegistryTest.java [hunk 16]
```diff
@@ -0,0 +1,347 @@
...
+        Timer timer = Timer.builder(METER_NAME)
+            .description(METER_DESCRIPTION)
+            .tags(Tags.of(meterTag))
+            .register(registry);
+        timer.record(10, TimeUnit.MILLISECONDS);
+        timer.record(77, TimeUnit.MILLISECONDS);
+        timer.record(111, TimeUnit.MILLISECONDS);
+
+        // This is where TimeWindowMax can be painful and make no sense at all. Need to
+        // re-visit this. Probably StepMax is what might fit good for OTLP.
```
Comment (shakuzen) [inline implementations/micrometer-registry-otlp/src/test/java/io/micrometer/registry/otlp/DeltaOtlpMeterRegistryTest.java:110]: I think we need a different Max implementation as you say. TimeWindowMax doesn't fit the interval as it should for OTLP.

**Review by shakuzen — COMMENTED** [review]
There are some remaining known issues, but if everyone thinks it would still be beneficial to get this in the RC1 (scheduled for Monday) for wider/easier testing while we fix the remaining issues before GA, I'm okay with that.

@lenin-jaganathan has already pointed out the issues in comments throughout the pull request:
* partial publish on close fix is not applied to this registry with delta temporality (maybe we can do something in the TCK to catch this proactively in registry implementations)
* time window max is not right for OTLP delta
* time window histogram is not right for OTLP delta

Depending on what we do about #2818 that may also cause issues here.

**Comment by lenin-jaganathan** [comment]
@shakuzen I do have a couple of POCs for StepBased Histogram and Max. But I just want this to look like a huge PR with too many new things. I would have a PR for StepHistogram and StepMax immediately after this PR. But If you think differently, I can just pile more commit on this. No preferences there, just a matter of getting Delta Registry support in to start with.

**Comment by shakuzen** [comment]
Sounds good. Please do follow-up with PRs for the other things. On the partial step issue, is there any reason we can't make `OtlpMeterRegistry` extend `StepMeterRegistry` to solve that?

**Comment by lenin-jaganathan** [comment]
@shakuzen Follow-up PR for to address Max and histogram - https://github.com/micrometer-metrics/micrometer/pull/3749

Let me just get back on partial publishing. I see an issue with partial publishing in some other Step Registries too (see [SignalFxTimer](https://github.com/micrometer-metrics/micrometer/blob/main/implementations/micrometer-registry-signalfx/src/main/java/io/micrometer/signalfx/SignalfxTimer.java) which was made an AbstractTimer just to have a custom Histogram implementation and now since it is no more Step the issue still exists.) And the same may be the case for OTLP as my new PR makes the OTLP Timer extend AbstarctTimer.

**Comment by shakuzen** [comment]
Yeah, it's on my list of things to do post-RC1 to more thoroughly check specific-registry implementations for this kind of thing. But it's expected if you make something not extend a Step meter, you're indicating you want to be responsible for any of the behavior that the step meters provide, such as the behavior on registry close. We can't really do anything about that. It depends on each registry implementation what's best for it, but I have to imagine it would be easier to extend and customize StepTimer than not in most cases.

**Comment by lenin-jaganathan** [comment]
I completely agree with the flexibility offered here. But since StepMeter is the interface that defines that we can perform a report with a different value on close, and this being package-private makes it someone cannot implement this behavior. I would think about something that lets implementers give an implementation for this last-minute publish but still let the StepRegistry close that.

> is there any reason we can't make OtlpMeterRegistry extend StepMeterRegistry to solve that
Technically, we do that since we almost override everything in the StepRegistry in OTLP. But, doing something like that will make me assume that OTLP is a StepRegistry which it is not as it can be either Step or Cumulative.
